File: languages/java/app-encryption/src/test/java/com/godaddy/asherah/crypto/BasicExpiringCryptoPolicyTest.java
package com.godaddy.asherah.crypto;

import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.concurrent.TimeUnit;

import static org.junit.jupiter.api.Assertions.*;

class BasicExpiringCryptoPolicyTest {
  private final static int testExpirationDays = 2;
  private final static int testCachingPeriod = 30;

  BasicExpiringCryptoPolicy policy = BasicExpiringCryptoPolicy
      .newBuilder()
      .withKeyExpirationDays(testExpirationDays)
      .withRevokeCheckMinutes(testCachingPeriod)
      .build();

  @Test
  void testIsKeyExpired() {
    Instant now = Instant.now();
    Instant before = now.minus(3, ChronoUnit.DAYS);

    assertTrue(policy.isKeyExpired(before));
  }

  @Test
  void testKeyIsNotExpired() {
    Instant now = Instant.now();
    Instant before = now.minus(1, ChronoUnit.DAYS);

    assertFalse(policy.isKeyExpired(before));
  }


  @Test
  void testRevokeCheckMillis() {
    assertEquals(TimeUnit.MINUTES.toMillis(testCachingPeriod), policy.getRevokeCheckPeriodMillis());
  }

  @Test
  void testDefaultsDontChange() {
    assertTrue(policy.canCacheSystemKeys());
    assertTrue(policy.canCacheIntermediateKeys());
    assertFalse(policy.canCacheSessions());
    assertEquals(1000, policy.getSessionCacheMaxSize());
    assertEquals(2 * 60 * 60 * 1000, policy.getSessionCacheExpireMillis());
    assertEquals(CryptoPolicy.KeyRotationStrategy.INLINE, policy.keyRotationStrategy());
    assertFalse(policy.notifyExpiredSystemKeyOnRead());
    assertFalse(policy.notifyExpiredIntermediateKeyOnRead());
  }

  @Test
  void testPrimaryBuilderPath() {
    BasicExpiringCryptoPolicy.KeyExpirationDaysStep builder = BasicExpiringCryptoPolicy.newBuilder();

    CryptoPolicy policy =
        builder.withKeyExpirationDays(testExpirationDays).withRevokeCheckMinutes(testCachingPeriod).build();
    assertNotNull(policy);
  }

  @Test
  void testFullBuilderPath() {
    BasicExpiringCryptoPolicy.KeyExpirationDaysStep builder = BasicExpiringCryptoPolicy.newBuilder();

    CryptoPolicy policy = builder
        .withKeyExpirationDays(testExpirationDays)
        .withRevokeCheckMinutes(testCachingPeriod)
        .withRotationStrategy(CryptoPolicy.KeyRotationStrategy.QUEUED)
        .withCanCacheSystemKeys(false)
        .withCanCacheIntermediateKeys(false)
        .withCanCacheSessions(true)
        .withSessionCacheMaxSize(42)
        .withSessionCacheExpireMinutes(33)
        .withNotifyExpiredSystemKeyOnRead(true)
        .withNotifyExpiredIntermediateKeyOnRead(true)
        .build();

    assertNotNull(policy);
    assertEquals(CryptoPolicy.KeyRotationStrategy.QUEUED, policy.keyRotationStrategy());
    assertFalse(policy.canCacheSystemKeys());
    assertFalse(policy.canCacheIntermediateKeys());
    assertTrue(policy.canCacheSessions());
    assertEquals(42, policy.getSessionCacheMaxSize());
    assertEquals(TimeUnit.MINUTES.toMillis(33), policy.getSessionCacheExpireMillis());
    assertTrue(policy.notifyExpiredSystemKeyOnRead());
    assertTrue(policy.notifyExpiredIntermediateKeyOnRead());
  }

}

File: languages/csharp/AppEncryption/AppEncryption.Tests/Crypto/BasicExpiringCryptoPolicyTest.cs
using System;
using GoDaddy.Asherah.Crypto;
using Xunit;

namespace GoDaddy.Asherah.AppEncryption.Tests.Crypto
{
    [Collection("Logger Fixture collection")]
    public class BasicExpiringCryptoPolicyTest
    {
        private static readonly int TestExpirationDays = 2;
        private static readonly int TestCachingPeriod = 30;

        private readonly BasicExpiringCryptoPolicy policy = BasicExpiringCryptoPolicy.NewBuilder()
            .WithKeyExpirationDays(TestExpirationDays)
            .WithRevokeCheckMinutes(TestCachingPeriod)
            .Build();

        [Fact]
        private void TestIsKeyExpired()
        {
            DateTimeOffset now = DateTimeOffset.UtcNow;
            DateTimeOffset before = now.AddDays(-3);

            Assert.True(policy.IsKeyExpired(before));
        }

        [Fact]
        private void TestKeyIsNotExpired()
        {
            DateTimeOffset now = DateTimeOffset.UtcNow;
            DateTimeOffset before = now.AddDays(-1);

            Assert.False(policy.IsKeyExpired(before));
        }

        [Fact]
        private void TestRevokeCheckMillis()
        {
            Assert.Equal(
                TimeSpan.FromMinutes(TestCachingPeriod).TotalMilliseconds,
                policy.GetRevokeCheckPeriodMillis());
        }

        [Fact]
        private void TestDefaultsDontChange()
        {
            Assert.True(policy.CanCacheSystemKeys());
            Assert.True(policy.CanCacheIntermediateKeys());
            Assert.False(policy.CanCacheSessions());
            Assert.Equal(1000, policy.GetSessionCacheMaxSize());
            Assert.Equal(2 * 60 * 1000, policy.GetSessionCacheExpireMillis());
            Assert.Equal(CryptoPolicy.KeyRotationStrategy.Inline, policy.GetKeyRotationStrategy());
            Assert.False(policy.NotifyExpiredSystemKeyOnRead());
            Assert.False(policy.NotifyExpiredIntermediateKeyOnRead());
        }

        [Fact]
        private void TestPrimaryBuilderPath()
        {
            BasicExpiringCryptoPolicy.IKeyExpirationDaysStep builder = BasicExpiringCryptoPolicy.NewBuilder();

            BasicExpiringCryptoPolicy basicExpiringCryptoPolicy = builder.
                WithKeyExpirationDays(TestExpirationDays).
                WithRevokeCheckMinutes(TestCachingPeriod).Build();
            Assert.NotNull(basicExpiringCryptoPolicy);
        }

        [Fact]
        private void TestFullBuilderPath()
        {
            BasicExpiringCryptoPolicy.IKeyExpirationDaysStep builder = BasicExpiringCryptoPolicy.NewBuilder();

            BasicExpiringCryptoPolicy basicExpiringCryptoPolicy = builder
                .WithKeyExpirationDays(TestExpirationDays)
                .WithRevokeCheckMinutes(TestCachingPeriod)
                .WithRotationStrategy(CryptoPolicy.KeyRotationStrategy.Queued)
                .WithCanCacheSystemKeys(false)
                .WithCanCacheIntermediateKeys(false)
                .WithCanCacheSessions(true)
                .WithSessionCacheMaxSize(42)
                .WithSessionCacheExpireMillis(33000)
                .WithNotifyExpiredSystemKeyOnRead(true)
                .WithNotifyExpiredIntermediateKeyOnRead(true)
                .Build();

            Assert.NotNull(basicExpiringCryptoPolicy);
            Assert.Equal(CryptoPolicy.KeyRotationStrategy.Queued, basicExpiringCryptoPolicy.GetKeyRotationStrategy());
            Assert.False(basicExpiringCryptoPolicy.CanCacheSystemKeys());
            Assert.False(basicExpiringCryptoPolicy.CanCacheIntermediateKeys());
            Assert.True(basicExpiringCryptoPolicy.CanCacheSessions());
            Assert.Equal(42, basicExpiringCryptoPolicy.GetSessionCacheMaxSize());
            Assert.Equal(33000, basicExpiringCryptoPolicy.GetSessionCacheExpireMillis());
            Assert.True(basicExpiringCryptoPolicy.NotifyExpiredSystemKeyOnRead());
            Assert.True(basicExpiringCryptoPolicy.NotifyExpiredIntermediateKeyOnRead());
        }
    }
}
